### PR TITLE
i.sentinel.download: add Earth Explorer as data-source for Sentinel-2 Level-1C data

### PR DESCRIPTION
**Pull request by griembauer**
# overview
This PR adds the option to download Sentinel-2 Level-1C data from USGS Earth Explorer to `i.sentinel.download`. It is based on the Python library [landsatxplore](https://github.com/yannforget/landsatxplore) and assumes that a minor fix in landsatxplore is included (https://github.com/yannforget/landsatxplore/pull/31).
Additionally, `i.sentinel.parallel.download` is adapted to these changes.

# usage

- the option `datasource` is added to indicate whether to search/download data from the Copernicus Open Access Hub or the USGS-Earth Explorer
- searching and downloading from USGS Earth Explorer works just as for the Copernicus Hub, but it is restricted to Sentinel-2 Level-1C data. File- and ID-naming differs for USGS Earth Explorer, as indicated in the first two columns: 
```
i.sentinel.download  start=2017-09-01 end=2017-12-01 clouds=20 producttype=S2MSI1C settings=usgs_credentials.txt output=some_folder datasource=USGS_EE -l
13 Sentinel product(s) found
5594498 L1C_T32ULB_A003180_20171015T104525 2017-10-15  0%
5501052 L1C_T31UGS_A003223_20171018T105026 2017-10-18  0%
5500630 L1C_T32ULB_A003223_20171018T105026 2017-10-18  0%
2810713 L1C_T32ULB_A003652_20171117T105314 2017-11-17  0%
[...]
```
- Users can also use the `uuid`-option to download a specific scene:
```
i.sentinel.download uuid=1448199,980971 settings=usgs_credentials.txt output=some_folder datasource=USGS_EE -l
980971 L1C_T31UGS_A008442_20170202T104536 2017-02-02 28%
1448199 L1C_T31UGS_A009686_20170430T103024 2017-04-30 18%
```
- Users can also use the `query` option with the ESA-specific keywords `identifier` or `filename`, and pass the keyword in **ESA-format**. This has the advantage that data can be searched from the ESA Hub (e.g. via `i.sentinel.coverage`) and downloaded directly from USGS:
```
i.sentinel.download  producttype=S2MSI1C query="identifier=S2A_MSIL1C_20170202T104241_N0204_R008_T32ULB_20170202T104536" settings=usgs_credentials.txt output=some_folder datasource=USGS_EE -l 
1 Sentinel product(s) found
979050 L1C_T32ULB_A008442_20170202T104536 2017-02-02 34%
```
- Finally, users can also search for/download scenes via the USGS-`identifier` using the `query`-option with the `usgs_identifier`-keyword:
```
i.sentinel.download  producttype=S2MSI1C query="usgs_identifier=L1C_T32ULB_A009014_20170314T104411" settings=usgs_credentials.txt output=some_dir  datasource=USGS_EE -l 
1193764 L1C_T32ULB_A009014_20170314T104411 2017-03-14 19%
```

- `i.sentinel.parallel.download` is adapted accordingly. If the user wants to use ESA-style identifiers/filenames but download from USGS, the `-e` flag can be activated.

# implementation
New methods were added to the `SentinelDownloader`-Class, or existing methods were extended. If `i.sentinel.download` should be extended by more datasources in the future, it might be helpful to create classes for each individual datasource and/or outsorce them to a new library for clarity. 

Translating the ESA-identifier to an USGS-identifier is not possible without further information, because the USGS-identifier contains the **absolute** orbit number (ESA only the relative). 
Getting the USGS-identifier from an ESA-identifier is now implemented as follows:

- Get acquisition date and S2-UTM-Tile ID from ESA-identifier
- Get the center coordinate of the UTM Tile from a kml containing all S-2 UTM-tiles.
- Search USGS_EE for Sentinel-2 data from this acquisition date, covering the center coordinate of the UTM tile
- Remove any found scenes that do not correspond to the required UTM tile (by identifier checking)

# known issues

- Occasionally, the USGS Earth Explorer does not respond when trying to download Sentinel-2 data. This happens for both the browser client (click on download --> new tab opens and keeps loading, fails eventually) and the API (results in `landsatxplore` error message after some waiting time). This does not correspond to specific datasets, as sometimes retrying can be successful. At the moment, it is not clear what's the cause of the problem and how to avoid it. 
- USGS Earth Explorer does **not** host **all** Sentinel-2 Level-1C datasets that can be found on the ESA Hub. This seems to apply especially for UTM-tiles with large overlap that come from the same scene. Often, there is just one tile available on USGS Earth Explorer while two can be found on the ESA Hub.

### CODE REVIEW

**Comment by lucadelu** [comment]
I had no way to test, but the code looks good.

**Comment by griembauer** [comment]
Update: 
It is now possible to download a specific scene using the `query="identifier=..."` or `query="filename=..."` option, using the ESA-filename convention. This makes it possible to use `i.sentinel.coverage` to identify scenes (footprints are **not** in the LTA) and then download them from USGS.

It is not possible to extract the USGS-sceneID from the ESA-filename because USGS uses the absolute orbit in the ID. One way would be to do another EsaHub-request to get the absolute orbit for a specific scene, but this would require having both an ESA-Hub and USGS-Session, and thus two login credentials at the same time. 

Instead it is implemented as follows: 

- Get acquisition date and S2-UTM-Tile ID from ESA-filename
- Get the center coordinate of the UTM Tile from the [kml](https://sentinel.esa.int/documents/247904/1955685/S2A_OPER_GIP_TILPAR_MPC__20151209T095117_V20150622T000000_21000101T000000_B00.kml/ec05e22c-a2bc-4a13-9e84-02d5257b09a8).
- Search USGS_EE for Sentinel-2 data from this acquisition date, covering the center coordinate of the UTM tile
- Remove any found scenes that do not correspond the the required UTM tile (by filename checking)

To Do:

- improve/shorten methods
- restrict use of footprints to ESA-Hub or find a way to get them from USGS
- ...

**Comment by griembauer** [comment]
Update: 
`i.sentinel.download` with option `datasource=USGS_EE` can now handle:

- option `uuid`: One or more ID in USGS-format (also called `entityID` in EE, but left as `uuid` for consistency) can be given to `i.sentinel.download`, any other filters will be ignored. USGS IDs have a format of six to seven digits and can be identified by e.g. running `i.sentinel.download -l` with the regular filters beforehand
- option `query`: The query option accepts now the keywords `query="identifier=..."`, `query="filename=..."` or `query=usgs_identifier=...`. The first two options allow the S2-identifier/filename in ESA-format (e.g. `S2A_MSIL1C_20180822T155901_N0206_R097_T17SPV_20180822T212023`). This enables to use the existing ESA-based functionality (e.g. `i.sentinel.coverage`) and do the downloading via USGS without the need to "translate" between identifiers. `query=usgs_identifier=...` allows an identifier in USGS-format, e.g. `L1C_T32ULB_A009686_20170430T103024`, which can also be identified by running `i.sentinel.download -l` with the regular filters beforehand

### example
Searching for scenes
```
i.sentinel.download  start=2017-09-01 end=2017-12-01 clouds=20 producttype=S2MSI1C settings=/home/griembauer/tmp/usgs_download/usgs_credentials output=/home/griembauer/tmp/usgs_download/files3 datasource=USGS_EE -l
13 Sentinel product(s) found
5594498 L1C_T32ULB_A003180_20171015T104525 2017-10-15  0%
5501052 L1C_T31UGS_A003223_20171018T105026 2017-10-18  0%
5500630 L1C_T32ULB_A003223_20171018T105026 2017-10-18  0%
2810713 L1C_T32ULB_A003652_20171117T105314 2017-11-17  0%
5594297 L1C_T31UGS_A003180_20171015T104525 2017-10-15  0%
2202140 L1C_T31UGS_A011831_20170927T103018 2017-09-27  1%
2303471 L1C_T32ULB_A012117_20171017T103024 2017-10-17  1%
2303156 L1C_T31UGS_A012117_20171017T103024 2017-10-17  2%
2201744 L1C_T32ULB_A011831_20170927T103018 2017-09-27  2%
2814365 L1C_T31UGS_A003652_20171117T105314 2017-11-17  7%
2126190 L1C_T31UGS_A011488_20170903T105515 2017-09-03 12%
2322853 L1C_T32ULB_A012160_20171020T104323 2017-10-20 16%
2325165 L1C_T31UGS_A012160_20171020T104323 2017-10-20 19%
```
Searching for USGS-style uuids:
```
i.sentinel.download uuid=1448199,980971 settings=/home/griembauer/tmp/usgs_download/usgs_credentials output=/home/griembauer/tmp/usgs_download/files3 datasource=USGS_EE -l
980971 L1C_T31UGS_A008442_20170202T104536 2017-02-02 28%
1448199 L1C_T31UGS_A009686_20170430T103024 2017-04-30 18%
```
Searching for identifier in ESA format:
```
i.sentinel.download  producttype=S2MSI1C query="identifier=S2A_MSIL1C_20170202T104241_N0204_R008_T32ULB_20170202T104536" settings=/home/griembauer/tmp/usgs_download/usgs_credentials output=/home/griembauer/tmp/usgs_download/files3 datasource=USGS_EE -l 
1 Sentinel product(s) found
979050 L1C_T32ULB_A008442_20170202T104536 2017-02-02 34%
```
Searching for identifier in USGS format
```
i.sentinel.download  producttype=S2MSI1C query="usgs_identifier=L1C_T32ULB_A009014_20170314T104411" settings=/home/griembauer/tmp/usgs_download/usgs_credentials  datasource=USGS_EE -l 
1193764 L1C_T32ULB_A009014_20170314T104411 2017-03-14 19%
```

**Comment by neteler** [comment]
Great improvement!

I tried it (for that I applied https://github.com/yannforget/landsatxplore/pull/31 locally).

Testing in an UTM zone 32N location:

```
GRASS : > g.proj -w
PROJCS["ETRS89 / UTM zone 32N",
    GEOGCS["ETRS89",
...
    AUTHORITY["EPSG","25832"]]

GRASS : > g.region vector=bonn_boundary_osm res=10 -a

GRASS : > i.sentinel.download start=2017-09-01 end=2017-12-01 clouds=20 producttype=S2MSI1C settings=~/.usgs output=~/tmp/usgs_s2/ datasource=USGS_EE -l
20 Sentinel product(s) found
5303824 L1C_T39NVJ_A003335_20171026T070624 2017-10-26  0%
2231308 L1C_T39NVH_A011929_20171004T071403 2017-10-04  0%
2328335 L1C_T39NVJ_A012215_20171024T071416 2017-10-24  0%
...
5480762 L1C_T39NVH_A003235_20171019T070516 2017-10-19  9%
2326596 L1C_T39NVJ_A012172_20171021T070353 2017-10-21 10%
2039313 L1C_T39NVJ_A011457_20170901T065934 2017-09-01 11%
```

Just the download fails:

```
GRASS : > i.sentinel.download uuid=2029698,2284220 settings=~/.usgs output=~/tmp/usgs_s2/ datasource=USGS_EE
ERROR: Unable to connect to USGS_EE: list indices must be integers or
       slices, not str
```

What I am wondering: The result looks like UTM zone 39 (?). Since I cannot download yet I could not check further.

For comparison, the ESA Hub result of the same region (of course with a different time range):

```
GRASS : > i.sentinel.download start=2020-09-01 end=2021-02-01 clouds=20 producttype=S2MSI1C settings=~/.sentinel_esa output=~/tmp/esa_s2/ datasource=ESA_COAH -l21 Sentinel product(s) found
1f20fb04-c4db-4c7a-9eaa-16066cf5eb5d S2B_MSIL1C_20200919T103649_N0209_R008_T32ULB_20200919T125309 2020-09-19T10:36:49Z  0% S2MSI1C
496ba07d-ec95-4f67-ad97-171d1b33dc2f S2B_MSIL1C_20200919T103649_N0209_R008_T31UGS_20200919T125309 2020-09-19T10:36:49Z  0% S2MSI1C
4f9eed68-7c8e-47fa-a178-735f829e73a7 S2A_MSIL1C_20200921T103031_N0209_R108_T31UGS_20200921T142406 2020-09-21T10:30:31Z  0% S2MSI1C
...
346f8ce0-73bd-44da-95f8-5c4f3a63b42c S2B_MSIL1C_20201128T104409_N0209_R008_T31UGS_20201128T114115 2020-11-28T10:44:09Z 15% S2MSI1C
66872f01-5701-4ffa-8c5b-b13ba123ef05 S2B_MSIL1C_20201218T104349_N0209_R008_T32ULB_20201218T115054 2020-12-18T10:43:49Z 18% S2MSI1C
```

**Comment by griembauer** [comment]
Hi @neteler thanks for testing! This specific uuid-download worked for me: 
```
i.sentinel.download uuid=2029698,2284220 settings=.usgs_credentials output=~/tmp/usgs_download datasource=USGS_EE
2 Sentinel product(s) found
Downloading data into </home/griembauer/tmp/usgs_download/files4>...
Downloading L1C_T30NZL_A012074_20171014T103459...
```
However I do get the same error when I use the master of [landsatxplore](https://github.com/yannforget/landsatxplore). It seems that there was a problem applying https://github.com/yannforget/landsatxplore/pull/31 locally. Could you try again by
- checking out the forked repo on the new branch (as long as the PR in landsatxplore is still open): https://github.com/griembauer/landsatxplore/tree/metadata_fix
- running `sudo python3 setup.py build` and `sudo python3 setup.py install`

**Comment by neteler** [comment]
I found the missing step: after cloning your "metadata_fix" fork, one needs to run prior to the installation `git checkout metadata_fix`. Now the download works:

```
GRASS : > i.sentinel.download uuid=2029698,2284220 settings=~/.usgs output=~/tmp/usgs_s2/ datasource=USGS_EE 
2 Sentinel product(s) found
Downloading data into </home/mneteler/tmp/usgs_s2/>...
Downloading L1C_T30NZL_A012074_20171014T103459...
13.7MB [00:16, 865kB/s]                                                                                                                 
Downloaded to
</home/mneteler/tmp/usgs_s2/S2A_MSIL1C_20171014T102021_N0205_R065_T30NZL_20171014T103459.SAFE>
Downloading L1C_T30NZL_A011502_20170904T103734...
12.6MB [00:31, 423kB/s]                                                                                                                 
Downloaded to
</home/mneteler/tmp/usgs_s2/S2A_MSIL1C_20170904T102021_N0205_R065_T30NZL_20170904T103734.SAFE>
```
Content:
```
...
    ├── GRANULE
    │   └── L1C_T30NZL_A012074_20171014T103459
    │       ├── AUX_DATA
    │       │   └── AUX_ECMWFT
    │       ├── IMG_DATA
    │       │   ├── T30NZL_20171014T102021_B01.jp2
    │       │   ├── T30NZL_20171014T102021_B02.jp2
    │       │   ├── T30NZL_20171014T102021_B03.jp2
    │       │   ├── T30NZL_20171014T102021_B04.jp2
    │       │   ├── T30NZL_20171014T102021_B05.jp2
    │       │   ├── T30NZL_20171014T102021_B06.jp2
    │       │   ├── T30NZL_20171014T102021_B07.jp2
    │       │   ├── T30NZL_20171014T102021_B08.jp2
    │       │   ├── T30NZL_20171014T102021_B09.jp2
    │       │   ├── T30NZL_20171014T102021_B10.jp2
    │       │   ├── T30NZL_20171014T102021_B11.jp2
    │       │   ├── T30NZL_20171014T102021_B12.jp2
    │       │   ├── T30NZL_20171014T102021_B8A.jp2
    │       │   └── T30NZL_20171014T102021_TCI.jp2
...
```
Nice!

TODO: the UTM zone as the retrieved S2 tile isn't in Bonn but covering [Null Island](https://en.wikipedia.org/wiki/Null_Island), hopefully just my mistake. Ah, I did not define `map=mymap`, maybe that's why? I kind of expected it would use the current region then since the description states "If not given then current computational extent is used". Seems yet to be failing when using the "USGS mode"?

**Comment by griembauer** [comment]
It is still weird that the results you receive are not in your current region. There is no need to use the `map` parameter, as without it, the extent of the current region is used. Could you share the exact location/mapset/region settings that lead to this result?

**Comment by neteler** [comment]
> It is still weird that the results you receive are not in your current region. There is no need to use the `map` parameter, as without it, the extent of the current region is used. Could you share the exact location/mapset/region settings that lead to this result?

Yes sure:

```
# EPSG 25833
GRASS utm33n_25833/rostock:~ > g.region -p
projection: 1 (UTM)
zone:       33
datum:      etrs89
ellipsoid:  grs80
north:      6014350
south:      5993070
west:       304090
east:       323570
nsres:      10
ewres:      10
rows:       2128
cols:       1948
cells:      4145344

g.version -rge
version=7.9.dev
date=2021
revision=8c8e5bce5
build_date=2021-02-04
build_platform=x86_64-pc-linux-gnu
build_off_t_size=8
libgis_revision=94cca3323
libgis_date=2021-02-04T07:32:23+00:00
proj=6.3.2
gdal=3.1.4
geos=3.8.1
sqlite=3.34.0

```

**Comment by griembauer** [comment]
Update: a timeout option was included to `landsatxplore` to avoid a) aborting downloads if EE responds slowly and b) waiting an infinite amount of time for a response: https://github.com/yannforget/landsatxplore/pull/32
`i.sentinel.download` now uses a conservative value of 10 minutes waiting time:
```
ee.download(scene_id=scene, output_dir=output, timeout=600)
```

**Comment by neteler** [comment]
I am testing with the new "landsatxplore v0.10", in the European morning, i.e. US late night:

```
GRASS nc_spm_08_grass7/user1:grass-addons > g.region -dp

# obtain scene list, from USGS server
i.sentinel.download datasource=USGS_EE start=2017-09-01 end=2017-12-01 clouds=10 producttype=S2MSI1C settings=~/.usgs -l
20 Sentinel product(s) found
2040545 L1C_T18STD_A011534_20170906T160933 2017-09-06  0%
2112375 L1C_T17SLU_A011620_20170912T163038 2017-09-12  0%
...
2115985 L1C_T18SVC_A011634_20170913T160335 2017-09-13  0%
2097358 L1C_T18STF_A011491_20170903T160146 2017-09-03  2%
2097933 L1C_T17SQA_A011491_20170903T160146 2017-09-03  4%
2115802 L1C_T18SUF_A011634_20170913T155123 2017-09-13  9%

# download first scene
i.sentinel.download datasource=USGS_EE start=2017-09-01 end=2017-12-01 clouds=10 producttype=S2MSI1C settings=~/.usgs  limit=1 output=~/tmp/usgs_s2/
1 Sentinel product(s) found
Downloading data into </home/mneteler/tmp/usgs_s2/>...
Downloading L1C_T18STF_A011491_20170903T160146...

...
(more traceback)
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/landsatxplore-0.10-py3.9.egg/landsatxplore/earthexplorer.py", line 76, in _download
    with self.session.get(url, stream=True, allow_redirects=True, timeout=timeout) as r:
  File "/usr/lib/python3.9/site-packages/requests/sessions.py", line 543, in get
    return self.request('GET', url, **kwargs)
  File "/usr/lib/python3.9/site-packages/requests/sessions.py", line 530, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/lib/python3.9/site-packages/requests/sessions.py", line 665, in send
    history = [resp for resp in gen]
  File "/usr/lib/python3.9/site-packages/requests/sessions.py", line 665, in <listcomp>
    history = [resp for resp in gen]
  File "/usr/lib/python3.9/site-packages/requests/sessions.py", line 237, in resolve_redirects
    resp = self.send(
  File "/usr/lib/python3.9/site-packages/requests/sessions.py", line 643, in send
    r = adapter.send(request, **kwargs)
  File "/usr/lib/python3.9/site-packages/requests/adapters.py", line 529, in send
    raise ReadTimeout(e, request=request)
requests.exceptions.ReadTimeout: HTTPSConnectionPool(host='dds.cr.usgs.gov', port=443): Read timed out. (read timeout=600)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/mneteler/.grass7/addons/scripts/i.sentinel.download", line 806, in <module>
    sys.exit(main())
  File "/home/mneteler/.grass7/addons/scripts/i.sentinel.download", line 798, in main
    downloader.download(options['output'], options['sleep'],
  File "/home/mneteler/.grass7/addons/scripts/i.sentinel.download", line 446, in download
    ee.download(scene_id=scene, output_dir=output, timeout=600)
  File "/usr/local/lib/python3.9/site-packages/landsatxplore-0.10-py3.9.egg/landsatxplore/earthexplorer.py", line 101, in download
    filename = self._download(url, output_dir, timeout=timeout)
  File "/usr/local/lib/python3.9/site-packages/landsatxplore-0.10-py3.9.egg/landsatxplore/earthexplorer.py", line 88, in _download
    raise EarthExplorerError(
landsatxplore.exceptions.EarthExplorerError: Connection timeout after 600 seconds.
```

Small wish: Would it be possible to try/except the timeout in the `ee.download()` call?

Meanwhile I'll try again later today (i.e. US daytime, maybe it matters).

**Comment by neteler** [comment]
Success (in US daytime):

```
i.sentinel.download datasource=USGS_EE start=2017-09-01 end=2017-12-01 clouds=10 producttype=S2MSI1C settings=~/.usgs  limit=1 output=~/tmp/usgs_s2/
1 Sentinel product(s) found
Downloading data into </home/mneteler/tmp/usgs_s2/>...
Downloading L1C_T18STF_A011491_20170903T160146...
171MB [00:23, 7.52MB/s]                                                                                                                                                                  
Downloaded to
</home/mneteler/tmp/usgs_s2/S2A_MSIL1C_20170903T160131_N0205_R054_T18STF_20170903T160146.SAFE>
```

**Comment by griembauer** [comment]
Update: 
The `timeout`-error is now wrapped in `try`/`except` to only show the final EarthExplorerError and not the entire traceback.

File: grass7/imagery/i.sentinel/i.sentinel.parallel.download/i.sentinel.parallel.download.py [hunk 1]
```diff
@@ -85,6 +85,23 @@
 #% answer: 1
 #%end
 
+#%option
+#% key: datasource
+#% description: Data-Hub to download scenes from.
```
Comment (neteler) [inline grass7/imagery/i.sentinel/i.sentinel.parallel.download/i.sentinel.parallel.download.py:90]:
```suggestion
#% description: Data-Hub to download scenes from
```

**Comment by neteler** [comment]
Please add yourself as an co-author, @griembauer  :)

**Review by neteler — APPROVED** [review]

**Comment by neteler** [comment]
May this PR be merged?

**Review by veroandreo — APPROVED** [review]

**Pull request merged**